xdem/coreg.py

```diff
@@ -2155,134 +2180,233 @@ def warp_dem(
 
-    :returns: a tuple containing 1) coregistered DEM as an xdem.DEM instance and 2) DataFrame of coregistration \
-statistics (count of obs, median and NMAD over stable terrain) before and after coreg.
+    :returns: A tuple containing 1) coregistered DEM as an xdem.DEM instance 2) the coregistration method \
+3) DataFrame of coregistration statistics (count of obs, median and NMAD over stable terrain) before and after \
+coregistration and 4) the inlier_mask used.
     """
-    # Check input arguments
-    if (coreg_method is not None) and ((hmode is not None) or (vmode is not None)):
-        warnings.warn("Both `coreg_method` and `hmode/vmode` are set. Using coreg_method.")
+    # Check inputs
+    if not isinstance(coreg_method, xdem.coreg.Coreg):
+        raise ValueError("`coreg_method` must be an xdem.coreg instance (e.g. xdem.coreg.NuthKaab())")
 
-    if hmode not in list(hmodes_dict.keys()):
-        raise ValueError(f"vhmode must be in {list(hmodes_dict.keys())}")
+    if isinstance(ref_dem_path, str):
+        if not isinstance(src_dem_path, str):
+            raise ValueError(
+                f"`ref_dem_path` is string but `src_dem_path` has type {type(src_dem_path)}."
+                "Both must have same type."
+            )
+    elif isinstance(ref_dem_path, gu.Raster):
+        if not isinstance(src_dem_path, gu.Raster):
+            raise ValueError(
+                f"`ref_dem_path` is of Raster type but `src_dem_path` has type {type(src_dem_path)}."
+                "Both must have same type."
+            )
+    else:
+        raise ValueError("`ref_dem_path` must be either a string or a Raster")
 
-    if vmode not in list(vmodes_dict.keys()):
-        raise ValueError(f"vmode must be in {list(vmodes_dict.keys())}")
+    if grid not in ["ref", "src"]:
+        raise ValueError(f"`grid` must be either 'ref' or 'src' - currently set to {grid}")
 
     # Load both DEMs
     if verbose:
         print("Loading and reprojecting input data")
-    if grid == "ref":
-        ref_dem, src_dem = gu.spatial_tools.load_multiple_rasters([ref_dem_path, src_dem_path], ref_grid=0)
-    elif grid == "src":
-        ref_dem, src_dem = gu.spatial_tools.load_multiple_rasters([ref_dem_path, src_dem_path], ref_grid=1)
+
+    if isinstance(ref_dem_path, str):
+        if grid == "ref":
+            ref_dem, src_dem = gu.spatial_tools.load_multiple_rasters([ref_dem_path, src_dem_path], ref_grid=0)
+        elif grid == "src":
+            ref_dem, src_dem = gu.spatial_tools.load_multiple_rasters([ref_dem_path, src_dem_path], ref_grid=1)
     else:
-        raise ValueError(f"`grid` must be either 'ref' or 'src' - currently set to {grid}")
+        ref_dem = ref_dem_path
+        src_dem = src_dem_path
+        if grid == "ref":
+            src_dem = src_dem.reproject(ref_dem, silent=True)
+        elif grid == "src":
+            ref_dem = ref_dem.reproject(src_dem, silent=True)
 
     # Convert to DEM instance with Float32 dtype
```

Comment:
I know this is already merged, but why do they have to be `float32` specifically? I get that integers are bad for coregistration, but I think this should at least be an argument. What if I'm working with float16 or float64?

Reply:
That's a good point! My intention was to avoid having int dtypes, but indeed, it doesn't have to be float32. On the other hand, it yields probably very similar results whether it is float16 or float64 and I feel that float32 is a good compromise.
I'll add a comment anyway to remind later us !